sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/MetricsBeamFnDataClient.java

```diff
@@ -133,50 +126,40 @@ public void drainAndBlock() throws Exception {
   }
 
   /**
-   * The QueueingFnDataReceiver is a a FnDataReceiver used by the QueueingBeamFnDataClient.
+   * The MetricsFnDataReceiver is a a FnDataReceiver used by the MetricsBeamFnDataClient.
    *
    * <p>All {@link #accept accept()ed} values will be put onto a synchronous queue which will cause
-   * the calling thread to block until {@link QueueingBeamFnDataClient#drainAndBlock} is called.
-   * {@link QueueingBeamFnDataClient#drainAndBlock} is responsible for processing values from the
+   * the calling thread to block until {@link MetricsBeamFnDataClient#drainAndBlock} is called.
+   * {@link MetricsBeamFnDataClient#drainAndBlock} is responsible for processing values from the
    * queue.
    */
-  public class QueueingFnDataReceiver<T> implements FnDataReceiver<WindowedValue<T>> {
+  public class MetricsFnDataReceiver<T> implements FnDataReceiver<WindowedValue<T>> {
     private final FnDataReceiver<WindowedValue<T>> consumer;
-    public InboundDataClient inboundDataClient;
 
-    public QueueingFnDataReceiver(FnDataReceiver<WindowedValue<T>> consumer) {
+    public MetricsFnDataReceiver(FnDataReceiver<WindowedValue<T>> consumer) {
       this.consumer = consumer;
     }
 
     /**
      * This method is thread safe, we expect multiple threads to call this, passing in data when new
-     * data arrives via the QueueingBeamFnDataClient's mainClient.
+     * data arrives via the MetricsBeamFnDataClient's mainClient.
+     *
+     * <p>This code is invoked CONCURRENTLY, by separate threads.
      */
     @Override
     public void accept(WindowedValue<T> value) throws Exception {
       try {
-        ConsumerAndData offering = new ConsumerAndData(this.consumer, value);
-        while (!queue.offer(offering, 200, TimeUnit.MILLISECONDS)) {
-          if (inboundDataClient.isDone()) {
-            // If it was cancelled by the consuming side of the queue.
-            break;
-          }
+        try (Closeable close = MetricsContainerStepMapEnvironment.setupMetricEnvironment()) {
```

Comment:
This is the major change, repurposing this class so that when elements come in on the gprc channel, we create a new MetricsContainerStepMap and ExecutionStateTracker for each element as a thread local variable. Keep track of them in a ConcurrentLinkedQueue and then iterate over all the MetricsContainerStepMaps at the end of the bundle and merge them to get the MonitoringInfos.